doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,653 @@
+QUIC stream SSL objects can have different settings for this option.
+
+Not supported for non-QUIC SSL objects.
+
+#### `SSL_get_rpoll_descriptor`, `SSL_get_wpoll_descriptor`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| New       | Never         | No        | CS            |
+
+These functions output poll descriptors which can be used to determine
+when the QUIC state machine should next be ticked. `SSL_get_rpoll_descriptor` is
+relevant if `SSL_want_net_read` returns 1, and `SSL_get_wpoll_descriptor` is
+relevant if `SSL_want_net_write` returns 1.
+
+The implementation of these functions is a simple forward to
+`BIO_get_rpoll_descriptor` and `BIO_get_wpoll_descriptor` on the underlying
+network BIOs.
+
+#### `SSL_want_net_read`, `SSL_want_net_write`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| New       | Never         | No        | CS            |
+
+These calls return 1 if the QUIC state machine is interested in receiving
+further data from the network, or writing to the network, respectively. The
+return values of these calls should be used to determine which wakeup events
+should cause an application to call `SSL_tick`. These functions do not mutate
+any state, and their return values may change after a call to any SSL function
+other than `SSL_want_net_read`, `SSL_want_net_write`,
+`SSL_get_rpoll_descriptor`, `SSL_get_wpoll_descriptor` and
+`SSL_get_tick_timeout`.
+
+#### `SSL_want`, `SSL_want_read`, `SSL_want_write`
+
+The existing API `SSL_want`, and the macros defined in terms of it, are
+traditionally used to determine if the SSL state machine has exited in
+non-blocking mode due to a desire to read from or write to the underlying
+network BIO. However, this API is unsuitable for use with QUIC because the
+return value of `SSL_want` can only express one I/O direction at a time (read or
+write), not both. This call will not be implemented for QUIC (e.g. always
+returns `SSL_NOTHING`) and `SSL_want_net_read` and `SSL_want_net_write` will be
+used instead.
+
+TBD: Should these be implemented for non-QUIC SSL objects?
+
+#### `SSL_set_initial_peer_addr`, `SSL_get_initial_peer_addr`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| New       | Never         | No        | CS            |
+
+`SSL_set_initial_peer_addr` sets the initial L4 UDP peer address for an outgoing
+QUIC connection.
+
+The initial peer address may be autodetected if no peer address has already been
+set explicitly and the QUIC connection SSL object is provided with a
+`BIO_s_dgram` with a peer set.
+
+`SSL_set_initial_peer_addr` cannot be called after a connection is established.
+
+### Future APIs
+
+A custom poller interface may be provided in the future. For more information,
```

Comment:
This is only relevant for blocking IO right? An application may choose to poll the network itself now and do its own event loop - no need for a new API.

Reply:
Yes absolutely, if it only uses QUIC in non-blocking mode this never comes into play. It is just a way to customise the poller function our blocking I/O adaptation layer uses. It is also trivial for an application to use QUIC in non-blocking mode and build its own blocking I/O adaptation layer out of that, if it wanted to for some reason, so it's more of a convenience than anything.